### DISCUSSION TITLE
Understanding/visualizing the MIDI data

### DISCUSSION THREAD

**Discussion by FDelporte** [atsushieno/ktmidi#62] [General]
Is there a parser/method/... included in ktmidi which can be used to convert the Midi bytes into "human-readable" data. Like note G9, for duration X with velocity Y? 

if not, this is a very useful explanation on StackOverflow: 
https://stackoverflow.com/questions/29481090/explanation-of-midi-messages

**Comment by atsushieno** [comment]
Hi,

> Is there a parser/method/... included in ktmidi which can be used to convert the Midi bytes into "human-readable" data. Like note G9, for duration X with velocity Y?

Not something we provide as part of our library - that kind of musicality features might be useful to some apps and if someone wants to have it in ktmidi I would happily merge a PR.

I myself find it app specific and implement in app code (like [this](https://github.com/atsushieno/kmmk/blob/f1703060281293dbcd8c0f3dea32f0150e589244/common/src/commonMain/kotlin/dev/atsushieno/kmmk/Model.kt#L43C9-L43C18)), especially in terms of note lengths (how would we represent them? Are durations available from the event stream like SMF, or are they based on the actual monotonic time? etc..).

If it is the other way around e.g. Kotlin enums to the status byte, then we have a handful of those such as [`MidiChannelStatus`](https://github.com/atsushieno/ktmidi/blob/0fffa319601f3b656965281cadbaa8520f1083e0/ktmidi/src/commonMain/kotlin/dev/atsushieno/ktmidi/MidiConstants.kt#L41).

**Reply by FDelporte** [reply]
Ah the enums look great! Thanks that helps me a lot. Indeed how notes are handled are probably very application depending...

**Reply by atsushieno** [reply]
(Some correction - they are not really Kotlin enums; I avoided it so that any future addition to the value list will not break the API...!)
